Documentation/wip/testing-proposal.md

```diff
@@ -0,0 +1,66 @@
+## Upgrade tests
+
+Currently in Fission, there is no way to run tests between upgrades and check if the upgrade broke something. This proposal aims to take a shot at this and while we are at it, a small diversion into related topic of testing! Please share your feedback and comments.
+
+### Goals
+- Newer version of Fission CLI should work with older version of Fission
+- A newer version of Fission passes all tests from previous version of Fission (Regression)
+- A cluster when upgraded from previous release behaves in the same way as when a newer version is installed from scratch on a cluster
+
+### Non-goals
+- Tests from newer version of Fission working with older versions of Fission
+
+### Proposal
+
+The upgrade test is run only once in a certain frequency (Mostly once a day) based on a cron job. Two version of fission source code are checked out in two subdirectories, the parameters for the version could be specified as environment variables.
+
+```
+./fission
+./fission_old
+```
+
+1. Fission_old is built and deployed in the cluster and then tests are run. This is exact same as the current build and test workflow but for a previous version of Fission.
+
+2. Once (1) passes, script switches to fission directory and builds and pushes the images. The part where the workflow diverts is install_and_test, and here are the differences:
+
+- Instead of ```helm install```, we will run ```helm upgrade``` - and upgrade the existing cluster with newer images
+- Script will use the new fission CLI
+- Script will use tests in fission_old to run the tests
+
+3. The script for (1) can be used from existing codebase while the part (2) will need developing a part of script specifically for upgrade.
+
+4. After the run - if both set of tests succeed, the upgrade can be considered fairly successful (Although not entirely - as there can be potential gap)
+
+### Alternative approach considered
+
+It is possible to get the binary from previous release instead of source code - but it has two drawbacks:
```

Comment:
The runtime is much smaller, though.  Our tests are already ~20 minutes. You'll get to avoid building fission-bundle, fetcher, builder, cli, etc.